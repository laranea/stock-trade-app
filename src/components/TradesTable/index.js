// @flow
import React from 'react';

import type { Trade } from '../../types/tradeTypes';
import Widget from '../Widget';
import Table from '../Table';

type TradesTableUi = {
  price: number,
  quantity: number,
  stockSymbol: string,
  timestamp: string
};

type TradesTableProps = {
  tradesData: Array<Trade>,
  handleGetTrades: () => void
};

const colDefs = ['Transaction Date', 'Stock', 'Price', 'No. Shares Purchased'];

export const formatData = (data: Array<Trade>): Array<TradesTableUi> => {
  return (
    data
      .map(item => ({
        timestamp: item.timestamp.toLocaleString(),
        stockSymbol: item.stockSymbol,
        price: item.price,
        quantity: item.quantity
      }))
      // @DESC - sort the trades in reverse chronological order.
      .sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp))
  );
};

export default class TradesTable extends React.Component<TradesTableProps> {
  componentDidMount(): void {
    this.props.handleGetTrades();
  }

  render() {
    return (
      <Widget title="Recent Trades">
        <Table colDefs={colDefs} rowData={formatData(this.props.tradesData)} />
      </Widget>
    );
  }
}
