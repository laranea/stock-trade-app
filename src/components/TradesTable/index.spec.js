import React from 'react';
import { shallow } from 'enzyme';

import TradesTable, { formatData } from './';

const MOCK_DATA = [
  {
    id: 'ABC12345',
    price: 100,
    quantity: 200,
    stockSymbol: 'TEST',
    timestamp: new Date(
      'Sat May 04 2019 17:47:10 GMT+0100 (British Summer Time)'
    )
  },
  {
    id: 'ABC12345',
    price: 100,
    quantity: 200,
    stockSymbol: 'TEST',
    timestamp: new Date(
      'Sun May 05 2019 17:47:10 GMT+0100 (British Summer Time)'
    )
  },
  {
    id: 'ABC12345',
    price: 100,
    quantity: 200,
    stockSymbol: 'TEST',
    timestamp: new Date(
      'Sun May 05 2019 12:12:10 GMT+0100 (British Summer Time)'
    )
  }
];

const MOCK_FORMATTED_DATA = [
  {
    price: MOCK_DATA[1].price,
    quantity: MOCK_DATA[1].quantity,
    stockSymbol: MOCK_DATA[1].stockSymbol,
    timestamp: MOCK_DATA[1].timestamp.toLocaleString()
  },
  {
    price: MOCK_DATA[2].price,
    quantity: MOCK_DATA[2].quantity,
    stockSymbol: MOCK_DATA[2].stockSymbol,
    timestamp: MOCK_DATA[2].timestamp.toLocaleString()
  },
  {
    price: MOCK_DATA[0].price,
    quantity: MOCK_DATA[0].quantity,
    stockSymbol: MOCK_DATA[0].stockSymbol,
    timestamp: MOCK_DATA[0].timestamp.toLocaleString()
  }
];

const handleGetTradesMock = jest.fn();

describe('TradesTable component', () => {
  it('should match snapshot', () => {
    const component = shallow(
      <TradesTable
        tradesData={MOCK_DATA}
        handleGetTrades={handleGetTradesMock}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should call handleGetTrades on load', () => {
    shallow(
      <TradesTable
        tradesData={MOCK_DATA}
        handleGetTrades={handleGetTradesMock}
      />
    );
    expect(handleGetTradesMock).toHaveBeenCalled();
  });

  describe('When calling formattedData', () => {
    it('should return formatted data', () => {
      const result = formatData(MOCK_DATA);
      expect(result).toEqual(MOCK_FORMATTED_DATA);
    });
  });
});
