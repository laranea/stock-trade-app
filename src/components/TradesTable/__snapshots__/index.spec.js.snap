// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradesTable component should match snapshot 1`] = `
<Widget
  title="Recent Trades"
>
  <TableComponent
    colDefs={
      Array [
        "Transaction Date",
        "Stock",
        "Price",
        "No. Shares Purchased",
      ]
    }
    rowData={
      Array [
        Object {
          "price": 100,
          "quantity": 200,
          "stockSymbol": "TEST",
          "timestamp": "5/5/2019, 5:47:10 PM",
        },
        Object {
          "price": 100,
          "quantity": 200,
          "stockSymbol": "TEST",
          "timestamp": "5/5/2019, 12:12:10 PM",
        },
        Object {
          "price": 100,
          "quantity": 200,
          "stockSymbol": "TEST",
          "timestamp": "5/4/2019, 5:47:10 PM",
        },
      ]
    }
  />
</Widget>
`;
